### PR TITLE
Error if unprotected header contains crit param

### LINKED ISSUE #80: “crit” label is permitted in unprotected headers
Per the [COSE spec](https://datatracker.ietf.org/doc/html/rfc8152#section-3.1), 

> When present, this parameter MUST be placed in the protected header bucket.

Currently, messages with `HeaderLabelCritical ` in `UnprotectedHeaders` are marshaled and unmarshaled without error. The implementation should validate that the value is found only in the Protected headers.

- From Trail of Bits audit report

### PR DESCRIPTION
**Pull request by qmuntal**
Fixes #80

Signed-off-by: qmuntal <qmuntaldiaz@microsoft.com>

### CODE REVIEW

File: headers.go [hunk 1]
```diff
@@ -401,6 +407,16 @@ func ensureHeaderIV(h map[interface{}]interface{}) error {
 	return nil
 }
 
+// ensureNoCritical ensures crit parameter is not present in the header.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-3.1
+func ensureNoCritical(h map[interface{}]interface{}) error {
+	if hasLabel(h, HeaderLabelCritical) {
+		return errors.New("unprotected header: crit parameter must be places in the protected header")
```
Comment (thomas-fossati) [inline headers.go:415]: nit: `h` is just a header map from the point of view of this function.  The calling site should add the rest of the context.

```suggestion
		return errors.New("unexpected crit parameter found")
```

File: headers.go [hunk 2]
```diff
@@ -211,6 +214,9 @@ func (h *UnprotectedHeader) UnmarshalCBOR(data []byte) error {
 	if err := decMode.Unmarshal(data, &header); err != nil {
 		return err
 	}
+	if err := ensureNoCritical(header); err != nil {
+		return err
```
Comment (thomas-fossati) [inline headers.go:218]: see also suggestion below

```suggestion
		return fmt.Errorf("unprotected header: %w")
```

File: headers.go [hunk 3]
```diff
@@ -182,6 +182,9 @@ func (h UnprotectedHeader) MarshalCBOR() ([]byte, error) {
 	if err := validateHeaderLabel(h); err != nil {
 		return nil, err
 	}
+	if err := ensureNoCritical(h); err != nil {
+		return nil, err
```
Comment (thomas-fossati) [inline headers.go:186]:
```suggestion
		return nil, fmt.Errorf("unprotected header: %w", err)
```

**Review by thomas-fossati — COMMENTED** [review]
Thanks!  I've inlined a couple of suggestions.

**Review by thomas-fossati — APPROVED** [review]
thanks!

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Comment by qmuntal** [comment]
Can we have this merged?

**Pull request merged**